dali/kernels/kernel_traits.h

```diff
@@ -0,0 +1,243 @@
+    typename filter<Args, is_kernel_arg>::type...
+  >;
+};
+
+template <typename Kernel>
+struct kernel_run_inputs {
+  using type = typename detail::input_lists<decltype(Kernel::Run)>::type;
+};
+
+
+template <typename Kernel>
+struct kernel_run_outputs {
+  using type = typename detail::output_lists<decltype(Kernel::Run)>::type;
+};
+
+
+template <typename Kernel>
+struct kernel_run_args {
+  using type = typename detail::kernel_arg_params<decltype(Kernel::Run)>::type;
+};
+
+
+IMPL_HAS_NESTED_TYPE(Inputs)
+IMPL_HAS_NESTED_TYPE(Outputs)
+IMPL_HAS_NESTED_TYPE(Args)
+
+template <typename Kernel, bool has_explicit_def = has_type_Inputs<Kernel>::value>
+struct KernelInputs {
+  using type = typename Kernel::Inputs;
+};
+
+template <typename Kernel>
+struct KernelInputs<Kernel, false> {
+  using type = typename kernel_run_inputs<Kernel>::type;
+};
+
+template <typename Kernel, bool has_explicit_def = has_type_Outputs<Kernel>::value>
+struct KernelOutputs {
+  using type = typename Kernel::Outputs;
+};
+
+template <typename Kernel>
+struct KernelOutputs<Kernel, false> {
+  using type = typename kernel_run_outputs<Kernel>::type;
+};
+
+template <typename Kernel, bool has_explicit_def = has_type_Args<Kernel>::value>
+struct KernelArgs {
+  using type = typename Kernel::Args;
+};
+
+template <typename Kernel>
+struct KernelArgs<Kernel, false> {
+  using type = typename kernel_run_args<Kernel>::type;
+};
+
+}  // namespace detail
+
+/// @brief Tells what inputs a kernel takes
+///
+/// If there's a type `Kernel::Inputs`, then this type is returned.
+/// Otherwise, it's a tuple of all `InList` parameters from `Kernel::Run` signature.
+template <typename Kernel>
+using kernel_inputs = typename detail::KernelInputs<Kernel>::type;
+
+/// @brief Tells what outputs a kernel produces
+///
+/// If there's a type `Kernel::Outputs`, then this type is returned.
+/// Otherwise, it's a tuple of all `OutList` parameters from `Kernel::Run` signature.
+template <typename Kernel>
+using kernel_outputs = typename detail::KernelOutputs<Kernel>::type;
+
+/// @brief Tells what extra arguments a kernel takes
+///
+/// If there's a type `Kernel::Args`, then this type is returned.
+/// Otherwise returns all parameters to `Kernel::Run` that are neither
+/// `InList`, `OutList` or KernelContext.
+template <typename Kernel>
+using kernel_args = typename detail::KernelArgs<Kernel>::type;
+
+namespace detail {
+
+IMPL_HAS_UNIQUE_FUNCTION(Run)
+IMPL_HAS_UNIQUE_FUNCTION(GetRequirements)
+
+template <typename Kernel>
+std::is_same<void, decltype(apply_all(Kernel::Run,
+    *(KernelContext*)nullptr,  // NOLINT
+    std::declval<kernel_outputs<Kernel>>(),
+    std::declval<kernel_inputs<Kernel>>(),
+    std::declval<kernel_args<Kernel>>() ))>
+  IsKernelRunnable(Kernel*);
+
+std::false_type IsKernelRunnable(...);
+
+template <typename Kernel>
+std::is_same<KernelRequirements, decltype(apply_all(Kernel::GetRequirements,
+    *static_cast<KernelContext*>(nullptr),  // NOLINT
```

Comment:
Is there a difference between this cast here, and the cast few lines above (the C-style one)?